### DISCUSSION TITLE
Custom `typeFn` for date strings?

### DISCUSSION THREAD

**Discussion by calvinwyoung** [nartc/mapper#384] [Q&A] [answered]
A common use case we have for automapper is converting `Date` objects into date strings (i.e., `YYYY-MM-DD`). Right now we're applying a `dateToStringConverter` to each date field as described [in the docs](https://automapperts.netlify.app/docs/mapping-configuration/convert-using#usage). This works fine, but is very verbose when working with a lot of date properties.

We'd love to be able to define a DTO like this:

```typescript
export class FooDto {
  @AutoMap({ typeFn: () => DateString })
  date?: DateString;
}
```

This could use a custom `DateString` type that could look something like this:

```typescript
export class DateString extends String {
  constructor(value: unknown) {
    super(formatAsDateString(value));
  }
}
```

(This specific implementation is admittedly gross — I'm just riffing on the `@AutoMap({ typeFn: () => String })` pattern that's common for string enums.)

Is there way to implement something similar with automapper? Ideally we'd like to use the same pattern to convert any type of non-JSON-serializable object to strings.

**Comment by micalevisk** [comment]
what about [custom type converters](https://automapperts.netlify.app/docs/mapping-configuration/custom-type-converter)?

**Comment by calvinwyoung** [comment]
The issue with that approach is that sometimes we want to map a `Date` object to an ISO timestamp string (e.g., `2000-01-01T01:00:00.000Z`) and other times we want to map it to a date string (e.g., `2000-01-01`). In both cases, the mapping is from `Date` -> `string`

Is there a way to implement a custom type converter that can distinguish between these two cases?

**Reply by micalevisk** [reply]
oh, got it.

> Is there a way to implement a custom type converter that can distinguish between these two cases?

I guess not.

**Comment by micalevisk** [comment, marked as answer]
can you show how you're using `ConvertUsing`?

**Reply by calvinwyoung** [reply]
Our use case looks roughly like this:

```typescript
export class Source {
  @AutoMap()
  fooDate?: Date;

  @AutoMap()
  barTimestamp?: Date;
}

export class Destination {
  @AutoMap()
  fooDate?: string;

  @AutoMap()
  barTimestamp?: string;
}


mapper
  .createMap(Source, Destination)

  // Convert `fooDate` using `dateToStringConverter` to output `YYYY-MM-DD`.
  .forMember(
    (destination) => destination.fooDate,
    convertUsing(dateToStringConverter, (source) => source.fooDate)
  )

  // Convert `barTimestamp` using `timestampToStringConverter` to output `YYYY-MM-DDTHH:mm:ss.sssZ`.
  .forMember(
    (destination) => destination.barTimestamp,
    convertUsing(timestampToStringConverter, (source) => source.barTimestamp)
  );  
```

I recognize that part of complexity is that Javascript doesn't have a primitive that distinguishes between dates and timestamps (e.g., the Python stdlib includes both `date` and `datetime` objects).

Is there a better solution that doesn't involve copying the `forMember` / `convertUsing` syntax for each property for each source-destination-pair?

**Reply by micalevisk** [reply]
I can't think of anything better than doing `.forMember(d => d.fooDate, withDateToString('fooDate'))`

---

The solution you're proposing is actually feasible with custom type converter but _"Type Converters only works with Primitive types at the moment"_, which is why you can't achieve that for now. I don't know why this limitation exists tbh.

**Reply by calvinwyoung** [reply]
Ah that's really unfortunate to hear. Thanks for the thoughtful responses — it's greatly appreciated 🙏

**Reply by nartc** [reply]
Thank you Micael for responding! Thanks Calvin for the issue! The limitation is because I doubted myself lol. I think I can make something work for your use-case.

**Reply by nartc** [reply]
How does this look?
```ts
export class DateString extends Date {}

export class TimestampString extends Date {}

export class TypeConverterSource {
  @AutoMap()
  value1!: string;
  @AutoMap()
  value2!: string;
  @AutoMap()
  value3!: string;
  @AutoMap({ typeFn: () => DateString })
  value4!: Date;
  @AutoMap({ typeFn: () => TimestampString })
  value5!: Date;
}

export class TypeConverterDestination {
  @AutoMap()
  value1!: number;
  @AutoMap()
  value2!: Date;
  @AutoMap()
  value3!: boolean;
  @AutoMap()
  value4!: string;
  @AutoMap()
  value5!: string;
}
```
I have 2 custom classes `DateString` and `TimestampString` which extend `Date`. In my example, they really do nothing except for differentiating between themselves. Then on `value4` and `value5` of `Source`, I let `AutoMapper` knows that: "Hey, `value4` is type `Date` for TypeScript, but for you Mapper, it is of type `DateString`." (Same thing applies to `value5`)

On the `Destination`, I set `value4` and `value5` to type `string`. This will allow me to setup the following `TypeConverter`:

```ts
mapper
   .addTypeConverter(DateString, String, (ds) => ds.toDateString())
   .addTypeConverter(TimestampString, String, (ts) => ts.toISOString())
   .createMap(TypeConverterSource, TypeConverterDestination);
```

This tells `AutoMapper` that: "Hey, when you see `DateString` that is being mapped to `String`, use `ds.toDateString()`. When you see `TimestampString` that is being mapped to `String`, use `ts.toISOString()`.

Here's a test case:

```ts
mapper.addProfile(typeConverterProfile);

const date = new DateString('10/14/1991');
const ts = new TimestampString('10/14/1991');

const source = new TypeConverterSource();
source.value4 = date;
source.value5 = ts;

const destination = mapper.map(
   source,
   TypeConverterDestination,
   TypeConverterSource
);
expect(destination.value4).toEqual(date.toDateString());
expect(destination.value5).toEqual(date.toISOString());
```
![image](https://user-images.githubusercontent.com/25516557/146705414-41002d87-0a31-494c-8d71-97f4fb0ade99.png)
cc @calvinwyoung

**Reply by calvinwyoung** [reply]
@nartc This is brilliant and solves our problem exactly. Thank you for the quick response, and for maintaining such a great library!

**Reply by nartc** [reply]
@calvinwyoung Cool! I'll cut a release tomorrow morning. It's getting late here. Thank you for the kind words

**Reply by calvinwyoung** [reply]
Amazing! Thanks so much!

**Reply by nartc** [reply]
@calvinwyoung 7.1 has been released. Please check it out. It might have edge-cases so if you run into issues, feel free to open them here.

**Reply by calvinwyoung** [reply]
Wow thanks for getting that released so quickly — we really appreciate it :)

It looks like this is *almost* resolved on our end. Since some of our date / timestamp properties are optional, our custom type converters need to handle optional values:

```typescript
mapper
  .addTypeConverter(DateString, String, (ds) => (ds ? someDateConversionFunction(ds) : undefined))
  .addTypeConverter(TimestampString, String, (ts) => (ts ? someTimestampConversionFunction(ts) : undefined));
```

Unfortunately this causes a type error since `undefined` can't be assigned to the destination type. What's your guidance for mixing type converters with optional values?

**Reply by nartc** [reply]
We can relax the return type of the converter function here to return `TDestination | undefined`. As of this moment, can you handle `undefined` in `someDateConversionFunction()` then use `non-null assertion` (hope that you don't have ESLint rule preventing this 😅). 

```ts
mapper
   .addTypeConverter(DateString, String, ds => someDateConversionFunction(ds)!)
```

**Reply by calvinwyoung** [reply]
Haha yes no problem, that's exactly what we're doing right now, and we'll wait for another release with relaxed type constraints.

Thanks again for being so responsive — automapper is becoming an increasingly core part of our Typescript stack :)

**Reply by calvinwyoung** [reply]
@nartc Happy new year! Just wanted to see if you had a sense for when you might be able to relax the return types for the converter function. We're still using the above workaround but it'd be nice to remove it if possible. Thank you!

**Reply by nartc** [reply]
@calvynwyoung Happy new year to you too! I'll release it tomorrow. Tbh, I forgot about this as it's marked Answered 😄

**Reply by calvinwyoung** [reply]
🤣  No problem at all! Thanks for the quick turnaround!

**Reply by nartc** [reply]
@calvinwyoung 7.2.1 has been released with relaxed type

**Reply by calvinwyoung** [reply]
This works beautifully! Thank you!!
